Move away from fedora 31 and use distroless/base instead

victortoso opened the pull request:
**What this PR does / why we need it**:

While creating a binary in my local Fedora 37 and pushing it to Fedora31 container, it would not run due conflict with libc. Fedora 31 was EOL in 2020-11-24, so the base image is overdue to an upgrade.

Further discussion in slack,  it was suggested to use distroless/base image as that is something we update more often.

**Release note**:
```release-note
NONE
```

jean-edouard reviewed (commented):
Can we bump the aarch64 image too? Thanks!

EdDev reviewed (commented):
Where do we use these images at the moment?

There are also other images that are fedora based, like `fedora_with_test_tooling` and `fedora_realtime` which run in the e2e tests.

victortoso commented:
> Where do we use these images at the moment?

I noticed it with the sidecar examples, seems to be mostly there:

```
toso@tapioca ~/s/k/kubevirt (bump-fedora-image)> grep -rniI "@fedora_aarch64\/\|@fedora\/" *
cmd/example-cloudinit-hook-sidecar/BUILD.bazel:37:        "@io_bazel_rules_go//go/platform:linux_arm64": "@fedora_aarch64//image",
cmd/example-cloudinit-hook-sidecar/BUILD.bazel:38:        "//conditions:default": "@fedora//image",
cmd/example-hook-sidecar/BUILD.bazel:39:        "@io_bazel_rules_go//go/platform:linux_arm64": "@fedora_aarch64//image",
cmd/example-hook-sidecar/BUILD.bazel:40:        "//conditions:default": "@fedora//image",
cmd/example-disk-mutation-hook-sidecar/BUILD.bazel:38:        "@io_bazel_rules_go//go/platform:linux_arm64": "@fedora_aarch64//image",
cmd/example-disk-mutation-hook-sidecar/BUILD.bazel:39:        "//conditions:default": "@fedora//image",
images/winrmcli/BUILD.bazel:17:        "@io_bazel_rules_go//go/platform:linux_arm64": "@fedora_aarch64//image",
images/winrmcli/BUILD.bazel:18:        "//conditions:default": "@fedora//image",
tests/BUILD.bazel:87:        "@io_bazel_rules_go//go/platform:linux_arm64": "@fedora_aarch64//image",
tests/BUILD.bazel:88:        "//conditions:default": "@fedora//image",
```

> There are also other images that are fedora based, like fedora_with_test_tooling and fedora_realtime which run in the e2e tests.

Yes, [kubevirtci/fedora-with-test-tooling](https://quay.io/repository/kubevirtci/fedora-with-test-tooling) has both `fedora_with_test_tooling` and `fedora_with_test_tooling_aarch64`. I think it shouldn't be a problem to use that instead. Agreed?

EdDev commented:
> Yes, [kubevirtci/fedora-with-test-tooling](https://quay.io/repository/kubevirtci/fedora-with-test-tooling) has both `fedora_with_test_tooling` and `fedora_with_test_tooling_aarch64`. I think it shouldn't be a problem to use that instead. Agreed?

Those have additional tooling and are modified to fit the testing needs of VM/s, but most importantly they are used as images for VM/s in our tests.
So I am not sure they can replace the ones you mentioned.

But what we can surely see is that we need a way to automate this, at least to detect we are not up to date.

I asked previously about who uses these images to understand if we actually exercise them somewhere.
I personally only know about `winrmcli` which is used in the windows (e2e) tests.

victortoso commented:
/retest

victortoso commented:
Failure of `pull-kubevirt-e2e-windows2016` might be a real issue as I've changed the base image for `winrmcli`
The test failures are all like:
```
20:59:43:   Timed out after 300.002s.
20:59:43:   Unexpected error:
20:59:43:       <*errors.errorString | 0xc001135cf0>: {
20:59:43:           s: "failed executing command on pod: unable to upgrade connection: container not found (\"winrmcli\"): stderr : stdout: ",
20:59:43:       }
20:59:43:       failed executing command on pod: unable to upgrade connection: container not found ("winrmcli"): stderr : stdout: 
20:59:43:   occurred
20:59:43:   In [It] at: tests/windows_test.go:177 
```

So, the issue seems related to the container not being deployed/found. Trying to run this tests locally, they are being skipped due `Skip tests that requires PV disk-windows` from [this check](https://github.com/kubevirt/kubevirt/blob/main/tests/framework/checks/skips.go#L134). I'm guessing I need to prepare the environment before running the test. I'll try tomorrow.

P.S: I did a quick test in the images locally and I can run the binary (without arguments, just to see if works)
```bash
toso@tapioca ~/s/k/kubevirt (main) > podman run --entrypoint= --tls-verify=false -it localhost:40641/kubevirt/winrmcli:devel /usr/bin/winrm-cli
ERROR: Please enter the command to execute on the command line

toso@tapioca ~/s/k/kubevirt (main) > podman run --entrypoint= --tls-verify=false -it localhost:40641/kubevirt/winrmcli:devel /usr/bin/winrmcp
Source directory is required.
```

victortoso commented:
/retest

victortoso commented:
JFI, I removed the change on `winrmcli` as I could not reproduce locally the error mentioned in https://github.com/kubevirt/kubevirt/pull/9229#issuecomment-1440935591 - Also, this image is used for tests so it might be a better fit to move and fetch it to `kubevirtci` like other images used for tests.

After this PR is merged, only `winrmcli` base image uses the old fedora image. Wdyt @EdDev ?

alicefr commented:
/cc

alicefr commented:
> Where do we use these images at the moment?
> 
> There are also other images that are fedora based, like `fedora_with_test_tooling` and `fedora_realtime` which run in the e2e tests.

@EdDev I don't quite understand why we want to consider those images. They are container-disks not container images. AFAIU, this PR aims to update the container based image of the sidecars, or am I missing something?

EdDev commented:
> @EdDev I don't quite understand why we want to consider those images. They are container-disks not container images. AFAIU, this PR aims to update the container based image of the sidecars, or am I missing something?

Well, I was originally unclear who uses the images that have been upgraded and just mentioned that there are other Fedora's in use except that.
There are images for sidecars but also for tooling and for e2e tests (e.g. that odd `winrmcli`). And beyond that we have the container-disks.

As long as it is mentioned what is effected by this bump, I think all is good.

victortoso commented:
Hi @EdDev,

> As long as it is mentioned what is effected by this bump, I think all is good.

I'm not bumping Fedora, instead I moved those to distroless/base image as that is something we update more often. I've updated the initial comment to make this clear. As mentioned in https://github.com/kubevirt/kubevirt/pull/9229#issuecomment-1446364025 only `winrmcli` is using Fedora 31.

Would this PR me enough as is or should I rework something?  `fossa` test does not like me, but I raised this in https://github.com/kubevirt/kubevirt/issues/9305

alicefr commented:
The `kubevirt-fossa` is already a while that is failing, and it isn't mandatory to merge PRs. So, don't worry to much about it. I think Daniel has a PR to fix it.

alicefr commented:
@victortoso do we want to include `winrmcli`? Have you been able to reproduce the error again in the CI?

victortoso commented:
> @victortoso do we want to include winrmcli? Have you been able to reproduce the error again in the CI?

@alicefr I've added the commit that changes winrmcli base image, plus the one that removes Fedora 31/32 from `WORKSPACE`. Soon we will be able to see `CI` artifacts again.

alicefr commented:
@victortoso sleep is missing in the container have a look [here](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/9229/pull-kubevirt-e2e-windows2016/1633475357870919680/artifacts/k8s-reporter/1_pods.log):
```sh
                        "name": "winrmcli",
                        "state": {
                            "waiting": {
                                "reason": "CreateContainerError",
                                "message": "container create failed: time=\"2023-03-08T14:51:08Z\" level=error msg=\"runc create failed: unable to start container process: exec: \\\"sleep\\\": executable file not found in $PATH\"\n"
                            }
                        },
```

alicefr commented:
Either we find a way to avoid sleep or we need to create an rpm tree for this image including sleep (this should be relatively simple)

alicefr commented:
Probably, replacing `sleep` with `tail -f /dev/null ` will be sufficient

victortoso commented:
> Probably, replacing `sleep` with `tail -f /dev/null ` will be sufficient: https://github.com/kubevirt/kubevirt/blob/main/tests/sysprep_test.go#L375

In fedora, `sleep` and `tail` are provided by the same `coreutils` package, so I wonder if `tail` will be available if `sleep` is not? Trying locally:
```bash
toso@tapioca ~/s/k/kubevirt (bump-fedora-image) [127]> podman run --entrypoint= --tls-verify=false -it localhost:40747/kubevirt/winrmcli:devel /usr/bin/sleep
Error: preparing container 55d4f4e6efda78b9f7c74ce852d07c58bb99ab0f30e73e8bef621b2e4f88e5f2 for attach: crun: executable file `/usr/bin/sleep` not found in $PATH: No such file or directory: OCI runtime attempted to invoke a command that was not found
toso@tapioca ~/s/k/kubevirt (bump-fedora-image) [127]> podman run --entrypoint= --tls-verify=false -it localhost:40747/kubevirt/winrmcli:devel /usr/bin/tail
Error: preparing container 2dcf1b0400ddf4614a98570b3273188d4d1f80e7073d9f12b8d0bb12f8cc0f9e for attach: crun: executable file `/usr/bin/tail` not found in $PATH: No such file or directory: OCI runtime attempted to invoke a command that was not found
```

victortoso commented:
> Well, then we need an rpm tree that adds it

While looking at this, I saw we do have a `coreutils-single-0__8.32-34.el9.x86_64` rpm definition in `WORKSPACE` with a rpmtree called `testimage_x86_64` that uses it and it is consumed by `kubevirt-testing-base` container image in [images/BUILD.bazel](https://github.com/kubevirt/kubevirt/blob/main/images/BUILD.bazel#L7). The base image of it is distroless/base too, but with a few rpms included which should fit with with `winrmcli`. Let's see if CI agrees.

victortoso commented:
/retest

victortoso commented:
@EdDev does this  looks fine to you as well?

alicefr commented:
/approve
/hold
@EdDev can you please have a second look to this PR and then we can unblock it?

EdDev commented:
> friendly ping to @EdDev ~ [#9229 (comment)](https://github.com/kubevirt/kubevirt/pull/9229#issuecomment-1484990074)

Sorry for the delay. Looks fine to me, thanks.

alicefr commented:
/retest-requiered
